001-collection-backend-dispatching.md

```diff
@@ -0,0 +1,165 @@
+        return self.fallback.read_json(*args, engine=cudf.read_json, **kwargs)
+
+    def read_orc(self, *args, **kwargs):
+        return dask_cudf.read_orc(*args, **kwargs)
+    ...
+```
+
+Once the `DaskBackendEntrypoint` subclass is defined, the new entrypoint can be declared in the library's `setup.py` file (specifying the class with a `"dask.backends"` entrypoint).
+
+Note that the `CudfBackendEntrypoint` example above selects `PandasBackendEntrypoint` as the fallback entrypoint class, but does not directly inherit from this reference class. This approach allows Dask to properly move data from the pandas fallback for any IO functions that lack cudf-specific definitions. If the cudf subclass were to directly inherit from `PandasBackendEntrypoint`, then "fallback" behavior would not result in data-movement or user warnings.
+
+
+## Implementation Details
+
+**Reference Implementation** (Draft):
+
+- [Dask Component](https://github.com/rjzamora/dask/tree/backend-class) 
+- [CuDF Component](https://github.com/rjzamora/cudf/tree/backend-class) ("cudf" entrypoint definition in `dask_cudf`)
+
+### Dispatching Functions
+
+As described above, we propose that all IO functions for a specific backend be defined within a single `DaskBackendEntrypoint` subclass. The only subclasses defined within the dask source code will be the default reference subclasses for numpy and pandas. These entrypoint classes will be defined in the `backends.py` file for each collection, and will define all dispatch-able IO functions.
+
+The actual dispatching of IO functions will require the definition of a new `BackendDispatch` class in `dask.utils`. In contrast to the existing `dask.utils.Dispatch` class, `BackendDispatch` will use a backend string label (e.g. "pandas") for registration and dispatching, and the dispatching logic will be implemented at the `__getattr__` level (rather than in `__call__`). More specifically, registered "keys" and "values" for the dispatch class will correspond to backend labels and `DaskBackendEntrypoint` subclasses, respectively. When some Dask-collection code calls something like `backend_dispatch.read_parquet`, dispatching logic will be used to return the appropriate `"read_parquet"` attribute for the current backend.
+
+In order to avoid moving numpy- or pandas-specific IO logic into `backends.py`, the existing IO functions will  be renamed to `*_pandas`, and referenced "in place". To insure that the real IO functions are still defined at the same absolute and relative paths, and that the original doc-strings are recognized, we can add a few lines below the `*_pandas` definition to direct the original function name to the dispatching machinery:
+
+
+```python
+def read_parquet_pandas(...):
+    <Previous read_parquet definition and doc-string>
+
+def read_parquet(*args, **kwargs):
+    return dataframe_backend_dispatch.read_parquet(*args, **kwargs)
+
+read_parquet.__doc__ = read_parquet_pandas.__doc__
```

Comment:
I'm interested in the docstrings. Often we inherit them from pandas and just augment them a bit, but maybe they should also be part of the dispatch mechanism.